x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/EnrichPolicy.java

```diff
@@ -85,6 +90,9 @@ public EnrichPolicy(StreamInput in) throws IOException {
             in.readString(),
             in.readStringList()
         );
+        if (in.readBoolean()) {
```

Comment:
and then:

```java
if (in.readBoolean()) {
   version = Version.readVersion(in);
} else {
   version = null;
}
```

?

Reply:
Is the idea here then to just make the policy object fully immutable, and when storing it make a copy with the version set before saving it to the cluster state? I think I can manage that.

Reply:
Yes, that works.
Or we can set version to Version.CURRENT if version is null in the 2nd constructor?

Reply:
I think this is what `Version.V_EMPTY` is for. 

+1 to remove boolean in favor of `V_EMPTY` if null.

Reply:
I'm not against setting it to Version.CURRENT if it would be `null` otherwise. Does it make sense that a user could create a policy with `version_created` already set to a version? Do we care to limit that?

Reply:
+1 to always set to Version.CURRENT if null.  

> Does it make sense that a user could create a policy with version_created already set to a version? 

nope

> Do we care to limit that?

Doesn't your [validation](https://github.com/elastic/elasticsearch/pull/45021/files#diff-af743c8fa51d5b94efadde37df70798fR66) already do that ?

Reply:
The validation will fail if we set the version to Version.CURRENT no matter what - If the user ends up not providing the version in the request, when the object is deserialized from XContent, the version will be null, which will mean it would get set to CURRENT, thus removing any way to tell if a user actually set the version or not.

Reply:
What if do this:

Add an argument to fromXContent:

```
public static EnrichPolicy fromXContent(XContentParser parser, boolean fromAPI) throws IOException {
    return PARSER.parse(parser, fromAPI);
}
```

Change parsers' context type from void to Boolean:

```
private static final ConstructingObjectParser<EnrichPolicy, Boolean> PARSER = new ConstructingObjectParser<>("policy",
```

Change parser:

```
parser.declareField(ConstructingObjectParser.optionalConstructorArg(),
            (p, fromApi) -> {
            if (fromApi) {
                // fail    
            }
            return Version.fromString(p.text());
            },
            VERSION_CREATED,
            ObjectParser.ValueType.STRING
        );
```

This way we fail if a version is provided via api and otherwise we're ok with a version being provided?

Reply:
btw I'm happy with the PR as is. I don't feel strong about this (^) any more.